### PR TITLE
Add invert_permute operator with SYCL implementation for XPU

### PR DESCRIPTION
**Pull request by aagalleg**
This PR introduces the `invert_permute` operator to fbgemm-xpu, enabling inverse permutations on Intel XPU devices. 

#### Core Implementation
- **SYCL Kernel** (`invert_permute_kernel.cpp/h`): High-performance SYCL implementation that computes inverse permutations on XPU devices
- **Operator Registration** (`ops_registry.cpp`): Registers the operator with PyTorch's dispatch system using conditional schema registration to avoid conflicts
- **Python API** (`ops.py`): Clean Python wrapper function with type hints for easy integration

#### Infrastructure
- **Build System** (`CMakeLists.txt`): Added SYCL kernel to build configuration
- **Torch Library Utilities** (`torch_library.h`): Helper utilities for schema existence checking and operator registration
- **Cleanup**: Removed `.gitkeep` files from populated directories

#### Testing
- **Comprehensive Test Suite** (`test_invert_permute.py`):
  - Correctness validation for int32/int64 data types
  - Edge cases (empty, single element, identity, reverse permutations)
  - Input validation and error handling
  - CPU-XPU consistency verification

cc: @manuelhsantana, @flezaalv

### CODE REVIEW

**Review by dvrogozh — COMMENTED** [review]
@aagalleg, I missed the beginning of the PRs chain and left all the comments in https://github.com/intel/torchlib-xpu/pull/73. Comments are pretty generic, can you please, refer to them?

For the convenience, please, mark all follow up PRs as drafts - in this way we will clearly know what should be reviewed first.

File: packages/fbgemm-xpu/src/fbgemm_xpu/fbgemm_utils/torch_library.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace fbgemm_xpu::utils::torch {
+
+inline bool schemaExists(const std::string& qualified_name) {
+  return c10::Dispatcher::singleton()
+      .findSchema({qualified_name, ""})
+      .has_value();
+}
+
+} // namespace fbgemm_xpu::utils::torch
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/fbgemm_utils/torch_library.h]: Add empty line in the end of file.

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/invert_permute_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+#include "invert_permute_kernel.h"
+
+namespace fbgemm_xpu {
+
+// ============================================================================
+// SYCL Kernel Functor Implementations
+// ============================================================================
+
+void InvertPermuteKernelInt32::operator()(const sycl::nd_item<1>& item) const {
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/invert_permute_kernel.cpp]: This seems templetable kernel over the operand type. Why don't we do that especially considering that the reference CUDA kernel is doing exactly that [1]?

[1] https://github.com/pytorch/FBGEMM/blob/8744183e4bd7390a674a0d8736bf4d7ef2d0516e/fbgemm_gpu/src/sparse_ops/sparse_invert_permute.cu#L15

File: packages/fbgemm-xpu/src/fbgemm_xpu/ops.py [hunk 3]
```diff
@@ -5,12 +5,13 @@
 # Python wrapper functions for all custom operators under the fbgemm namespace
 # This module provides user-friendly interfaces to the C++ operators
 
+import torch
+from torch import Tensor
+
 __all__ = [
-	"dense_embedding_codegen_lookup_function",
+    "invert_permute",
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/ops.py]: Here I am confused. Why do we expose these XPU specific APIs from fbgemm xpu? who is supposed to use them? My understanding was that we are just register low level entrypoints for XPU key so that standard fbgemm APIs start to work on XPU device. Shouldn't that be that user calls `fbgemm::invert_permute` for example?

**Review by dvrogozh — COMMENTED** [review]
In addition to inline comments:
1. Repository has lint CI - please, address identified issues.
2. What we implement and merge must be tested under CI. And IMHO doing that by engineers who implemented the code is the best variant rather than offloading that to dedicated CI engineers. Thus, please, modify `.github/workflows/ci.yml` and add a test section for FBGEMM following the pattern outlined for torchcodec (see link below). The test job for FBGEMM will be much simpler. Target only `bmg` for now is fine. Note that fbgemm wheel is already getting build by CI.

https://github.com/intel/torchlib-xpu/blob/d4230aae4220649c6b879e43e2af166942abeebf/.github/workflows/ci.yml#L199

**Comment by flezaalv** [comment]
> In addition to inline comments:
> 
> 1. Repository has lint CI - please, address identified issues.
> 2. What we implement and merge must be tested under CI. And IMHO doing that by engineers who implemented the code is the best variant rather than offloading that to dedicated CI engineers. Thus, please, modify `.github/workflows/ci.yml` and add a test section for FBGEMM following the pattern outlined for torchcodec (see link below). The test job for FBGEMM will be much simpler. Target only `bmg` for now is fine. Note that fbgemm wheel is already getting build by CI.
> 
> https://github.com/intel/torchlib-xpu/blob/d4230aae4220649c6b879e43e2af166942abeebf/.github/workflows/ci.yml#L199

Hi @dvrogozh, regarding the CI, we added it here: #86. Please take a look and let me know if you have any comments.

File: packages/fbgemm-xpu/src/fbgemm_xpu/fbgemm_utils/torch_library.h (hunk 1 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/fbgemm_utils/torch_library.h]: Done

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/invert_permute_kernel.cpp (hunk 2 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/invert_permute_kernel.cpp]: Code was changed to use template over the operand type.

File: packages/fbgemm-xpu/src/fbgemm_xpu/ops.py (hunk 3 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/ops.py]: We used ops.py to expose the API from the start and didn't remove it once we tested the integration with fbgemm-gpu-cpu package. 
ops.py file is no longer necessary and it is now removed.

**Comment by aagalleg** [comment]
> In addition to inline comments:
> 
> 1. Repository has lint CI - please, address identified issues.
> 2. What we implement and merge must be tested under CI. And IMHO doing that by engineers who implemented the code is the best variant rather than offloading that to dedicated CI engineers. Thus, please, modify `.github/workflows/ci.yml` and add a test section for FBGEMM following the pattern outlined for torchcodec (see link below). The test job for FBGEMM will be much simpler. Target only `bmg` for now is fine. Note that fbgemm wheel is already getting build by CI.
> 
> https://github.com/intel/torchlib-xpu/blob/d4230aae4220649c6b879e43e2af166942abeebf/.github/workflows/ci.yml#L199

1- Done.
2- Regarding testing. The custom test_invert_permute.py has been removed. Instead, we follow the same convention as torchcodec-xpu and created a patch to be applied to the upstream FBGEMM tests v1.7.0. The patch adapts fbgemm_gpu/test/sparse/misc_ops_test.py to run test invert_permute on XPU. See 0001-Add-XPU-support-to-fbgemm-tests.patch.
Once PR #86 is merged to main I'll add the test to the CI.

**Review by dvrogozh — CHANGES_REQUESTED** [review]
@aagalleg, this looks exactly as what I expect to see from the enabling stand point. The only missing part is actual integration into CI run here: https://github.com/intel/torchlib-xpu/blob/main/.github/workflows/ci.yml. Do you work on that?

**Comment by aagalleg** [comment]
> @aagalleg, this looks exactly as what I expect to see from the enabling stand point. The only missing part is actual integration into CI run here: https://github.com/intel/torchlib-xpu/blob/main/.github/workflows/ci.yml. Do you work on that?

@flezaalv was working on the CI integration in PR #86. He added a [fbgemm-ci.yml](https://github.com/intel/torchlib-xpu/pull/86/changes#diff-1fcf4e4281fbed74a78d4c465ef2554d0da5daeb6193d55d066b011db60d5605) file in the workflows directory, not directly to ci.yml.

**Comment by flezaalv** [comment]
> > @aagalleg, this looks exactly as what I expect to see from the enabling stand point. The only missing part is actual integration into CI run here: https://github.com/intel/torchlib-xpu/blob/main/.github/workflows/ci.yml. Do you work on that?
> 
> @flezaalv was working on the CI integration in PR #86. He added a [fbgemm-ci.yml](https://github.com/intel/torchlib-xpu/pull/86/changes#diff-1fcf4e4281fbed74a78d4c465ef2554d0da5daeb6193d55d066b011db60d5605) file in the workflows directory, not directly to ci.yml.

@dvrogozh I added a dedicated CI file because I think torchcodec-xpu and fbgemm-xpu packages will have different release times. Please let me know if you have any comments.

**Comment by dvrogozh** [comment]
> @dvrogozh I added a dedicated CI file because I think torchcodec-xpu and fbgemm-xpu packages will have different release times. Please let me know if you have any comments.

We have a complexity here as besides `torchcodec-xpu` and `fbgemm-xpu` we also have an umbrella package called `torchlib-xpu` which depends on other 2. As a result we can't actually decouple releases of `torchcodec-xpu` and `fbgemm-xpu` right away. While we maintain packages in current design paradigm I prefer to have a single `ci.yml` file which handles everything. I have an idea how to reduce burden on the testing - I will post a PR shortly. Hold on.

**Comment by aagalleg** [comment]
@dvrogozh Patch apply and testing were added to CI.

**Review by dvrogozh — COMMENTED** [review]
```
../../../../github/home/miniconda3/envs/test-env/lib/python3.10/site-packages/torch/xpu/__init__.py:251
  /github/home/miniconda3/envs/test-env/lib/python3.10/site-packages/torch/xpu/__init__.py:251: UserWarning: Can't initialize Level Zero Sysman
    return _enum_zes_device_infos(visible_mask)

../../../../github/home/miniconda3/envs/test-env/lib/python3.10/site-packages/torch/xpu/__init__.py:278
  /github/home/miniconda3/envs/test-env/lib/python3.10/site-packages/torch/xpu/__init__.py:278: UserWarning: XPU device count is zero! (Triggered internally at /__w/pytorch/pytorch/c10/xpu/XPUFunctions.cpp:117.)
    count = torch._C._xpu_getDeviceCount() if zes_count < 0 else zes_count
```

These warnings in the tests are related to XPU. I am especially worried about the 2nd one as it basically says that XPU is unavailable. With that I am confused considering that `torch.xpu.is_available()` was returning `True`.

**Comment by dvrogozh** [comment]
>  I am especially worried about the 2nd one as it basically says that XPU is unavailable. With that I am confused considering that `torch.xpu.is_available()` was returning True.

Same warning exists in TorchCodec CI as well: https://github.com/intel/torchlib-xpu/actions/runs/29928230040/job/88954373767.

**Comment by dvrogozh** [comment]
https://github.com/pytorch/pytorch/blob/18f0b90367333ef99381cb3cb2164085a4733f04/torch/xpu/__init__.py#L278

Ok, I think above 2 warnings are related. The XPU works for us because there is a fallback to detect XPU devices with SYCL runtime if detection with L0 has failed. I am not sure though why L0 path has failed - this seems to be an issue (not with torchlib).

**Comment by dvrogozh** [comment]
Issue with CI runner is tracked and closed in https://github.com/intel/torchlib-xpu/issues/104.

**Review by dvrogozh — APPROVED** [review]
LGTM
